src/commands/classified/external.rs

```diff
@@ -43,233 +47,163 @@ impl Decoder for LinesCodec {
-                                &arg.tag,
-                            )
-                        } else {
-                            ShellError::labeled_error(
-                                "$it needs string data",
-                                "given something else",
-                                self.name_tag.clone(),
-                            )
-                        }
-                    })
-                })
-                .collect::<Result<Vec<String>, ShellError>>()?;
+pub(crate) async fn run_external_command(
+    command: ExternalCommand,
+    context: &mut Context,
+    input: ClassifiedInputStream,
+    stream_next: StreamNext,
+) -> Result<ClassifiedInputStream, ShellError> {
+    let stdin = input.stdin;
+    let inputs: Vec<Value> = input.objects.into_vec().await;
+
+    trace!(target: "nu::run::external", "-> {}", command.name);
+    trace!(target: "nu::run::external", "inputs = {:?}", inputs);
+
+    let mut arg_string = format!("{}", command.name);
+    for arg in command.args.iter() {
+        arg_string.push_str(&arg);
+    }
 
-            let commands = input_strings.iter().map(|i| {
-                let args = self.args.iter().filter_map(|arg| {
-                    if arg.chars().all(|c| c.is_whitespace()) {
-                        None
+    let home_dir = dirs::home_dir();
+
+    trace!(target: "nu::run::external", "command = {:?}", command.name);
+
+    let mut process;
+    if arg_string.contains("$it") {
+        let input_strings = inputs
+            .iter()
+            .map(|i| {
+                i.as_string().map(|s| s.to_string()).map_err(|_| {
+                    let arg = command.args.iter().find(|arg| arg.contains("$it"));
+                    if let Some(arg) = arg {
+                        ShellError::labeled_error(
+                            "External $it needs string data",
+                            "given row instead of string data",
+                            &arg.tag,
+                        )
                     } else {
-                        // Let's also replace ~ as we shell out
-                        let arg = if let Some(ref home_dir) = home_dir {
-                            arg.replace("~", home_dir.to_str().unwrap())
-                        } else {
-                            arg.replace("~", "~")
-                        };
-
-                        Some(arg.replace("$it", &i))
+                        ShellError::labeled_error(
+                            "$it needs string data",
+                            "given something else",
+                            command.name_tag.clone(),
+                        )
                     }
-                });
-
-                format!("{} {}", self.name, itertools::join(args, " "))
-            });
+                })
+            })
+            .collect::<Result<Vec<String>, ShellError>>()?;
 
-            process = Exec::shell(itertools::join(commands, " && "))
-        } else {
```

Comment:
This feels a little surprising, but I'd need to try it out

Reply:
I'm not sure it was intentional. Are there tests for the original change?

Reply:
Not that I know of. Can we just use what was there before? We can focus on changing it in a more focused PR later.